event/event.go

```diff
@@ -8,27 +8,70 @@ import (
 	"go.uber.org/zap/zapcore"
 
 	uuid "github.com/satori/go.uuid"
+	"github.com/serverless/event-gateway/internal/zap"
+)
+
+
+const (
+	mimeJSON = "application/json"
+	mimeFormMultipart = "multipart/form-data"
+	mimeFormURLEncoded = "application/x-www-form-urlencoded"
 )
 
 // Event is a default event structure. All data that passes through the Event Gateway is formatted as an Event, based on
 // this schema.
 type Event struct {
-	Type       Type        `json:"event"`
-	ID         string      `json:"id"`
-	ReceivedAt uint64      `json:"receivedAt"`
-	Data       interface{} `json:"data"`
-	DataType   string      `json:"dataType"`
+	EventType          Type                   `json:"event-type" validate:"required"`
+	EventTypeVersion   string                 `json:"event-type-version"`
+	CloudEventsVersion string                 `json:"cloud-events-version" validate:"required"`
+	Source             string                 `json:"source" validate:"required"`
+	EventID            string                 `json:"event-id" validate:"required"`
+	EventTime          time.Time              `json:"event-time"`
+	SchemaURL          string                 `json:"schema-url"`
+	ContentType        string                 `json:"content-type"`
+	Extensions         zap.MapStringInterface `json:"extensions"`
+	Data               interface{}            `json:"data"`
 }
 
 // New return new instance of Event.
 func New(eventType Type, mime string, payload interface{}) *Event {
-	return &Event{
-		Type:       eventType,
-		ID:         uuid.NewV4().String(),
-		ReceivedAt: uint64(time.Now().UnixNano() / int64(time.Millisecond)),
-		DataType:   mime,
-		Data:       payload,
+	var event = &Event{
+		EventType:          eventType,
+		CloudEventsVersion: "0.1",
+		Source:             "https://slsgateway.com/",
+		EventID:            uuid.NewV4().String(),
+		EventTime:          time.Now(),
+		ContentType:        mime,
+		Data:               payload,
+	}
+
+	switch eventType {
+	case TypeInvoke, TypeHTTP:
+	default:
+		body, ok := payload.([]byte)
+		if !ok {
+			break
+		}
+		customevent := &Event{}
+		err := json.Unmarshal(body, customevent)
+		if err != nil || eventType != event.EventType {
+			break
+		}
+		event = customevent
+	}
+
+	// Because event.Data is []bytes here, it will be base64 encoded by default when being sent to remote function,
+	// which is why we change the event.Data type to "string" for forms, so that, it is left intact.
+	if eventbody, ok := event.Data.([]byte); ok && len(eventbody) > 0 {
+		switch {
+		case mime == mimeJSON:
+			json.Unmarshal(eventbody, &event.Data)
+		case strings.HasPrefix(mime, mimeFormMultipart), mime == mimeFormURLEncoded:
```

Comment:
You used `ParseMediaType` previously. Is it no longer needed?

Reply:
That is used in different PR. We'll import that when we merge, or do you want me to merge it now?

Reply:
Oh, sorry, yeah, let's leave it out here. We will merge it as a part of this second PR.